Include runtime errors in PEP8 checks and fix for various PEP8 statement errors

bsipocz opened the pull request:
[Edited]:

Fixes are for the followings, while only including the last two in ``setup.cfg``:

```
E711 - comparison to None should be 'if cond is None:'
E712 - comparison to True should be 'if cond is True:' or 'if cond:'
E713 - test for membership should be 'not in'
E714 - test for object identity should be 'is not'
E721 - do not compare types, use 'isinstance()'
E901 - SyntaxError or IndentationError
E902 - IOError
```

bsipocz commented:
Some of the changes caused local fails, but I thought it would be useful to see the results on the more comprehensive scale with running travis and then trying to figure out what to do.

bsipocz commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1049,14 +1049,14 @@ def test_assign_multiple_rows_to_table(self):
         assert (tbhdu2.columns.columns[3].array[4] ==
                 np.array([1., 2., 3., 4., 5.], dtype=np.float32)).all()
-        assert tbhdu2.columns.columns[4].array[4] == False
+        assert tbhdu2.columns.columns[4].array[4] is False
```
TIL about the diff between ``bool`` and ``numpy.bool_``. So the question is whether we want to:
 a) enforce E712, or 
 b) skip the few lines with pylint where both ``bool`` and ``numpy.bool_`` can be used, or
 c) convert one of them to the other

I would go with c) and have this here: ``assert bool(tbhdu2.columns.columns[4].array[4]) is False``

bsipocz commented:
@astrofrog - This is ready for review.

pllim commented on astropy/convolution/tests/test_kernel_class.py:
```diff
@@ -292,7 +292,7 @@ def test_custom_1D_kernel_list(self):
         Check if CustomKernel works with lists.
         """
         custom = CustomKernel([1, 1, 1, 1, 1])
-        assert custom.is_bool == True
```
Can `is_bool` be anything other than `True` or `False`? The following also does **not** raise `AssertionError`:

```python
assert 'mystring'
assert myclass
```

pllim replied:
I am wary of all removal of `== True` in this PR (across multiple sub-packages). Maybe it is safer to do `is True` unless you are very sure removing it will not change the context of the comparison.

bsipocz replied:
Yes, you're right. Here I just assumed blindly that is_bool should be bool (or maybe None).

mhvk replied:
Agree with @pllim that the test should be for `is True`, so that one automatically checks that it is a `Bool`

pllim commented on astropy/io/fits/column.py:
```diff
@@ -1095,7 +1095,7 @@ def _convert_to_valid_data_type(self, array):
                 # boolean needs to be scaled back to storage values ('T', 'F')
                 if array.dtype == np.dtype('bool'):
-                    return np.where(array == False, ord('F'), ord('T'))
+                    return np.where(array == np.bool(False), ord('F'), ord('T'))
```
I can't peg this change to any of the `Exxx` error listed in the PR.

bsipocz replied:
This was also E712 (``E712 comparison to False should be 'if cond is False:' or 'if not cond:'``), and was one of the examples where the original must be used to be able to do element-wise comparison. So to get rid of the error notification I could either add a pylint disable comment, or use this workaround.

pllim replied:
Since this is FITS, I am just going to tag @embray in case he has something to add...

mhvk replied:
I think here the comparison is correct as written, which would suggest a pylint disable ... ...  

Or, how about `array == np.False_`? Slightly more readable than `np.bool(False)

saimn replied:
`np.where(~array, ...)` should work also.

pllim commented on astropy/modeling/fitting.py:
```diff
@@ -897,7 +897,7 @@ def _fitter_to_model_params(model, fps):
     if has_tied:
         for idx, name in enumerate(model.param_names):
-            if model.tied[name] != False:
+            if model.tied[name]:
```
Similar to my comment about  removing `== True` above, perhaps it is safer to do `is not False` here.

mhvk replied:
I think here I like this better.

mhvk replied:
Like it better *in its changed form*

pllim commented on astropy/table/table.py:
```diff
@@ -675,7 +675,7 @@ def _convert_col_for_table(self, col):
         MaskedColumn if the table is masked.  Table subclasses like QTable
         override this method.
         """
-        if isinstance(col, Column) and not col.__class__ is self.ColumnClass:
```
Huh. Is the second comparison not equivalent to `not isinstance(col, self.ColumnClass)`?

mhvk replied:
Here, I think the test is right: one only wants to select `ColumnClass` subclass instances, not instances of `ColumnClass` itself. Aside: for speed, one probably wants to swap the two phrases `is not` is far faster than `isinstance` and usually `col.__class__` *is* `self.ColumnClass`.

pllim commented on astropy/time/tests/test_comparisons.py:
```diff
@@ -30,8 +30,8 @@ def test_miscompares(self):
                                      .format(op_str))
-        assert (t1 == None) is False
-        assert (t1 != None) is True
+        assert (t1 is None) is False
```
Isn't it simpler to be `assert t1 is not None`?

pllim commented on astropy/time/tests/test_comparisons.py:
```diff
@@ -30,8 +30,8 @@ def test_miscompares(self):
-        assert (t1 == None) is False
-        assert (t1 != None) is True
+        assert (t1 is None) is False
+        assert (t1 is not None) is True
```
This just looks like a duplicate to me. Maybe the original author can explain why both comparisons are needed.

mhvk replied:
That would be Tom A or me. Please keep these as `==` and `!=` --> they are specifically meant to test `Time.__eq__` and `Time.__ne__` -- and, no, we cannot count on everybody to stick to PEP8!
If you want to avoid disabling pylint, it would be fine to rewrite as
```
assert t1.__eq__(None) is False
```
etc. (though I think it is slightly less clear, and I dislike a bit rewriting tests just so PEP8 passes).

mhvk replied:
p.s. It may be good to write a comment above the test...

pllim commented on astropy/units/tests/test_units.py:
```diff
@@ -651,13 +651,13 @@ def test_compare_with_none():
     # because that doesn't trigger the bug.  See #3108.
-    assert not (u.m == None)
-    assert u.m != None
+    assert not (u.m is None)
```
The comment right above this one says, "We are deliberately not using `is None` here because that doesn't trigger the bug. See #3108."

bsipocz replied:
Thanks, my bad. Most of it was done with autopep8, and obviously my review wasn't thorough enough.

bsipocz commented:
@pllim - Thanks for the comments, I'll address them later today.

mhvk commented on astropy/convolution/tests/test_kernel_class.py:
```diff
@@ -201,7 +201,7 @@ def test_multiply_scalar_type(self, number):
         gauss = Gaussian1DKernel(3)
         gauss_new = number * gauss
-        assert type(gauss_new) == Gaussian1DKernel
+        assert isinstance(gauss_new, Gaussian1DKernel)
```
I think you need to keep the test here: we don't want the multiplication to suddenly start returning subclasses!  However, it should be
```
assert type(gauss_new) is Gaussian1DKernel
```

mhvk commented on astropy/convolution/tests/test_kernel_class.py:
```diff
@@ -210,7 +210,7 @@ def test_rmultiply_scalar_type(self, number):
         gauss = Gaussian1DKernel(3)
         gauss_new = gauss * number
-        assert type(gauss_new) == Gaussian1DKernel
+        assert isinstance(gauss_new, Gaussian1DKernel)
```
As above.

mhvk commented on astropy/io/fits/fitsrec.py:
```diff
@@ -419,7 +419,7 @@ def from_columns(cls, columns, nrows=0, fill=False):
                 # TODO: Maybe this step isn't necessary at all if _scale_back
                 # will handle it?
-                inarr = np.where(inarr == False, ord('F'), ord('T'))
+                inarr = np.where(inarr == np.bool(False), ord('F'), ord('T'))
```
Again, maybe use `np.False_`?

mhvk commented on astropy/io/fits/hdu/base.py:
```diff
@@ -888,7 +888,7 @@ def match_header(cls, header):
         # match_header for GroupsHDU will always be called before this one.
         if card.keyword == 'SIMPLE':
-            if 'GROUPS' not in header and card.value == False:
+            if 'GROUPS' not in header and card.value is False:
```
To keep this semantically the same, it would have to be `and not card.value`.  I think truthiness may be evaluated here based on value being empty, etc.

bsipocz replied:
@mhvk - I kept this as is, as raising an ``InvalidHDUException`` seems to be better for empty values.

mhvk commented on astropy/io/fits/hdu/compressed.py:
```diff
@@ -676,7 +676,7 @@ def match_header(cls, header):
             return False
 
-        if 'ZIMAGE' not in header or header['ZIMAGE'] != True:
+        if 'ZIMAGE' not in header or header['ZIMAGE'] is not True:
```
I would remove `!= True` (as you did above), i.e., let `header['ZIMAGE']` define truthiness itself.

mhvk commented on astropy/io/fits/hdu/groups.py:
```diff
@@ -503,7 +503,7 @@ def _verify(self, option='warn'):
         self.req_cards('GCOUNT', pos, _is_int, 1, option, errs)
         self.req_cards('PCOUNT', pos, _is_int, 0, option, errs)
-        self.req_cards('GROUPS', pos, lambda v: (v == True), True, option,
+        self.req_cards('GROUPS', pos, lambda v: (v), True, option,
```
This slightly worries me: what can `v` be? If it is guaranteed never to be an array, then this is OK (though parentheses can be removed). cc: @embray.

bsipocz replied:
My original changes made no sense, this should always return a bool, thus changed it to be ``is True``. And I'm pretty sure that the input is never an array (as being the value of the given header keyword).

mhvk commented on astropy/io/fits/hdu/hdulist.py:
```diff
@@ -629,7 +629,7 @@ def update_extend(self):
 
         if 'EXTEND' in hdr:
-            if len(self) > 1 and hdr['EXTEND'] == False:
+            if len(self) > 1 and hdr['EXTEND'] is False:
```
Again, I would suggest `not hdr['EXTEND']`.

mhvk commented on astropy/io/fits/hdu/image.py:
```diff
@@ -928,8 +928,8 @@ def __init__(self, data=None, header=None, do_not_scale_image_data=False,
         return (card.keyword == 'SIMPLE' and
-                ('GROUPS' not in header or header['GROUPS'] != True) and
-                card.value == True)
+                ('GROUPS' not in header or header['GROUPS'] is not True) and
```
And again...

bsipocz replied:
This was tricky, this one caused the regression in #5807, header['GROUPS'] is not a bool, and in the example case it's 2, so the second half of the expression becomes False rather than True

bsipocz replied:
Ohh, after having an actual look at the standard, this should have been OK as the value should have been boolean. 
My resolution is, never to make any stylistic change to the fits module again to avoid braking what's working fine.

mhvk commented on astropy/io/fits/hdu/image.py:
```diff
@@ -928,8 +928,8 @@ def __init__(self, data=None, header=None, do_not_scale_image_data=False,
-                ('GROUPS' not in header or header['GROUPS'] != True) and
-                card.value == True)
+                ('GROUPS' not in header or header['GROUPS'] is not True) and
+                card.value is True)
```
And again

mhvk commented on astropy/io/fits/tests/test_connect.py:
```diff
@@ -277,7 +277,7 @@ def test_bool_column(tmpdir):
 
     arr = np.ones(5, dtype=bool)
-    arr[::2] == False
+    arr[::2] is False
```
This definitely is wrong: maybe use `== np.False_` instead.

mhvk commented on astropy/io/fits/tests/test_division.py:
```diff
@@ -27,7 +27,7 @@ def test_card_with_continue(self):
     def test_valid_hdu_size(self):
         t1 = fits.open(self.data('tb.fits'))
-        assert type(t1[1].size) == type(1)
+        assert isinstance(t1[1].size, type(1))
```
I think here the test is for strict equality (though it should be `is` instead of `==`).

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -180,24 +180,24 @@ def test_assign_boolean(self):
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
+        assert h['BAR'] is False
```
I think here the test change is correct, but it should be `assert h['FOO'] is True`

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -180,24 +180,24 @@ def test_assign_boolean(self):
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
+        assert h['BAR'] is False
```
Here, I'm not sure what the actual intent was...  But testing it, it would seem that the idea is that any semi-boolean input gets converted to a proper `Bool`, so again `assert h['FOO'] is True`

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -180,24 +180,24 @@ def test_assign_boolean(self):
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
+        assert h['BAR'] is False
```
Idem (probably...)

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -1798,24 +1798,24 @@ def test_assign_boolean(self):
         h['BAR'] = False
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
```
`is True`

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -1798,24 +1798,24 @@ def test_assign_boolean(self):
         h['BAR'] = np.bool_(False)
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
```
`is True`

mhvk commented on astropy/io/fits/tests/test_header.py:
```diff
@@ -1798,24 +1798,24 @@ def test_assign_boolean(self):
         h.append(fits.Card.fromstring(barimg))
-        assert h['FOO'] == True
-        assert h['BAR'] == False
+        assert h['FOO']
```
`is True`

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -75,7 +75,7 @@ def comparerecords(a, b):
         if fieldb.dtype.char == 'S':
             fieldb = decode_ascii(fieldb)
-        if (type(fielda) != type(fieldb) and not
+        if (not isinstance(fielda, type(fieldb)) and not
```
I think this should be
```
if (not isinstance(fielda, type(fieldb)) and
     not isinstance(fieldb, type(fielda))):
```

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -976,7 +976,7 @@ def test_add_data_by_rows(self):
         assert (tbhdu.columns.columns[3].array[0] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu.columns.columns[4].array[0] == True
+        assert tbhdu.columns.columns[4].array[0]
```
`is True`

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1049,14 +1049,13 @@ def test_assign_multiple_rows_to_table(self):
                 np.array([1., 2., 3., 4., 5.], dtype=np.float32)).all()
-        assert tbhdu2.columns.columns[4].array[4] == False
-
+        assert bool(tbhdu2.columns.columns[4].array[4]) is False
```
This seems odd; if `array` elements are not `bool` already, I would just test with
```
assert not tbhdu2.columns.columns[4].array[4]
```

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1049,14 +1049,13 @@ def test_assign_multiple_rows_to_table(self):
         assert (tbhdu2.columns.columns[3].array[8] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu2.columns.columns[4].array[8] == False
+        assert bool(tbhdu2.columns.columns[4].array[8]) is False
```
Idem

mhvk commented on astropy/io/votable/tree.py:
```diff
@@ -2828,7 +2828,7 @@ def _write_binary(self, mode, w, **kwargs):
                     for i, converter in fields_basic:
                         try:
                             chunk = converter(array_row[i], array_mask[i])
-                            assert type(chunk) == type(b'')
```
Think this should be kept as is: presumably one doesn't want `chunk` to be a subclass of `Bytes`, so just `type(chunk) is type(b'')`.

bsipocz replied:
I think it's better to use ``six.binary_type`` here rather than then original version. Anyway I moved it back from doing isinstance.

mhvk commented on astropy/modeling/optimizers.py:
```diff
@@ -145,8 +145,8 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
         pars = [getattr(model, name) for name in model.param_names]
-        bounds = [par.bounds for par in pars if par.fixed != True and
-                  par.tied == False]
+        bounds = [par.bounds for par in pars if par.fixed is not True and
```
I think here using implicit truthiness is fine, i.e., `if not (par.fixed or par.tied)`

mhvk commented on astropy/modeling/tests/test_parameters.py:
```diff
@@ -1,4 +1,4 @@
-# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# Licensedf under a 3-clause BSD style license - see LICENSE.rst
```
Oops, typo!

mhvk commented on astropy/modeling/tests/test_parameters.py:
```diff
@@ -55,23 +55,23 @@ def test_parameter_properties():
+    assert p.fixed is False
     p.fixed = True
-    assert p.fixed == True
+    assert p.fixed
```
`is True` (will not repeat this any more; should have following @pllim!)

mhvk commented on astropy/table/table.py:
```diff
@@ -360,7 +360,7 @@ def __init__(self, data=None, masked=None, names=None, dtype=None,
         init_func(data, names, dtype, n_cols, copy)
 
         # Whatever happens above, the masked property should be set to a boolean
-        if type(self.masked) != bool:
```
`is bool` -- see comment above line.

mhvk commented on astropy/table/table.py:
```diff
@@ -2220,7 +2220,7 @@ def sort(self, keys=None):
                 raise ValueError("Table sort requires input keys or a table index")
             keys = [x.info.name for x in self.indices[0].columns]
-        if type(keys) is not list:
+        if not isinstance(keys, list):
```
Here, I think allowing subclasses is the right thing to do.

mhvk commented on astropy/table/tests/test_pickle.py:
```diff
@@ -72,11 +72,11 @@ def test_pickle_table(protocol):
-        assert type(tp['c']) == type(t['c'])
-        assert type(tp['d']) == type(t['d'])
-        assert type(tp['e']) == type(t['e'])
+        assert isinstance(tp['c'], type(t['c']))
```
In all these tests, one does want to be sure unpickling gives exactly the same class, so replace `==` by `is`

mhvk commented on astropy/table/tests/test_table.py:
```diff
@@ -260,7 +260,7 @@ def test_simple(self, table_types):
         assert np.all(t['a'].data == np.array([1, 2, 3]))
         assert np.all(t['b'].data == np.array([4, 5, 6], dtype=np.float32))
-        assert type(t['b'][1]) == np.float32
+        assert isinstance(t['b'][1], np.float32)
```
`==` -> `is`

mhvk commented on astropy/table/tests/test_table.py:
```diff
@@ -269,8 +269,8 @@ def test_from_np_array(self, table_types):
         assert np.all(t['b'] == np.array([4, 5, 6], dtype=np.float32))
-        assert type(t['a'][1]) == np.float64
-        assert type(t['b'][1]) == np.float32
+        assert isinstance(t['a'][1], np.float64)
```
Idem.

mhvk commented on astropy/units/tests/test_units.py:
```diff
@@ -623,7 +623,7 @@ def test_fractional_powers():
 
     x = u.cm ** Fraction(1, 2) * u.cm ** Fraction(2, 3)
-    assert type(x.powers[0]) == Fraction
+    assert isinstance(x.powers[0], Fraction)
```
`==` -> `is`

bsipocz replied:
The original version is full of this ``isinstance(x..., Fraction)`` checks (e.g. a few lines above). Do you think we should change all of those, too, or leave this as is to be consistent?

mhvk replied:
Looking at the context, I see your point. I think really either is fine, so maybe simplest is to just leave your change to `isinstance`.

mhvk commented on astropy/units/tests/test_units.py:
```diff
@@ -651,13 +651,13 @@ def test_compare_with_none():
 def test_validate_power_detect_fraction():
     frac = utils.validate_power(1.1666666666666665)
-    assert type(frac) == Fraction
+    assert isinstance(frac, Fraction)
```
`==` -> `is`

mhvk commented on astropy/utils/decorators.py:
```diff
@@ -123,7 +123,7 @@ def deprecated_func(*args, **kwargs):
         # This crazy way to get the type of a wrapper descriptor is
         # straight out of the Python 3.3 inspect module docs.
-        if type(func) != type(str.__dict__['__add__']):
+        if not isinstance(func, type(str.__dict__['__add__'])):
```
Given the comment above, I'd keep this as is...

mhvk commented on astropy/utils/decorators.py:
```diff
@@ -181,7 +181,7 @@ def deprecate(obj, message=message, name=name, alternative=alternative,
 
         altmessage = ''
-        if not message or type(message) == type(deprecate):
+        if not message or isinstance(message, type(deprecate)):
```
I'm not sure I understand decorators well enough, but my sense would be to leave this as a `==` -> `is` case; just don't see where subclasses could possibly come in.

mhvk commented on astropy/utils/decorators.py:
```diff
@@ -203,7 +203,7 @@ def deprecate(obj, message=message, name=name, alternative=alternative,
             return deprecate_function(obj, message)
 
-    if type(message) == type(deprecate):
+    if isinstance(message, type(deprecate)):
```
Idem

mhvk commented:
@bsipocz - very good idea, even if you have set yourself up for quite a bit of a chore here! Overall, my main comment would be that in tests one wants to be strict and thus check also that something is a `bool` (as said by @pllim), but in code one usually does not want to do that.

bsipocz commented:
@mhvk - Thanks for the review, and sorry that I left so many stuff in there. Now I only need to grab a beer and clearing this up will be actually quite pleasant :)

hamogu commented:
I've been outvoted on this several times in the past, but for the record:

[rant on]
I'm strongly opposed to enabling any more pep8 tests, style checks etc. Not because I don't like pep8 (I agree with most of what it says), but because each new test of this type raises the bar for new contributors. In python it's perfectly legal to say e.g. `type(a) == type(b)` and in most cases that will (i) do the job, (ii) be safe, and (iii) be perfectly readable for future generations.
Thus, I think we should accept that in astropy. Even as a fairly regular contributor I don't know all this stuff on top of my head; the learning curve for newcomers is a lot steeper (git, github, travis, [skip ci] ettiqutte, branching, whom to mention to get a review, how to rebase, pylint, sphinx, references in the docs, referenreces to intersphinx objects, p2 and p3 compatibility, ...). Serious contributors will be able to figure it out, but it's time consuming to learn all that stuff. 
We are in danger of loosing contributions from occasional contributors, e.g. Post-Docs who just want to contribute their useful statistics function, or their model fitter, or something, but who don't want to commit their live to coding.
[rant off]

astrofrog commented:
@bsipocz - I think that it might make more sense to tackle this in smaller PRs. What I would recommend is to make only a certain type of PEP8 change but NOT yet include them in the PEP8 tests. Then this does clean up the code while not yet raising the bar too high, until we decide to do so.

bsipocz commented:
Rebased and removed the E7* checks from the config file. Also fixed a few things merged to master since this PR was opened.

However kept the runtime error ones in there, as I think we should require all code committed to be able to run.

bsipocz commented:
This also includes #4980, otherwise there is a test failure.

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -976,7 +975,7 @@ def test_add_data_by_rows(self):
         assert (tbhdu.columns.columns[3].array[0] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu.columns.columns[4].array[0] == True
+        assert tbhdu.columns.columns[4].array[0] == True  # nopep8
```
Bit confused why the `== True` is not removed here, like below.

bsipocz replied:
There is a mixture of np.bool_ and bool types here where np.bool_ is not a subclass of bool, so it gave a failuer with ``is True``, though it wasn't clear to me where and why it became an np.True_ rather than a True. 
Thus left it unchanged.

mhvk replied:
I meant to just remove `== True` so that it just asserts truthiness, i.e.,
```
assert tbhdu.columns.columns[4].array[0]
```
you do that in all the lines below too, so I think is consistent.

bsipocz replied:
Ahh, I shouldn't have done it there either as per a previous comment (https://github.com/astropy/astropy/pull/4581/commits/df55f50e3c999085f589653ec24e868807b5d3d6#r52209960) and since the bool value is needed to be tested, too.

mhvk commented on astropy/table/tests/test_pickle.py:
```diff
@@ -72,11 +72,11 @@ def test_pickle_table(protocol):
-        assert type(tp['c']) == type(t['c'])
-        assert type(tp['d']) == type(t['d'])
-        assert type(tp['e']) == type(t['e'])
+        assert type(tp['c']) is type(t['c'])  # nopep8
```
Out of curiosity, what is wrong here from the pep8 perspective?

bsipocz replied:
not sure, it gives the same E721 warning as with ``==`` (``E721 do not compare types, use 'isinstance()'``), but it doesn't complain for simpler expression e.g. ``type(tp) is type(t)``

mhvk replied:
That seems weird, but o well, fine to just leave the comment.

mhvk commented on astropy/utils/decorators.py:
```diff
@@ -123,7 +123,7 @@ def deprecated_func(*args, **kwargs):
         # This crazy way to get the type of a wrapper descriptor is
         # straight out of the Python 3.3 inspect module docs.
-        if type(func) != type(str.__dict__['__add__']):
+        if type(func) != type(str.__dict__['__add__']):  # nopep8
```
Did this have a problem with `is not`?

bsipocz replied:
it still complains for E721, but you're right ``is not`` is still better.

mhvk commented on astropy/wcs/tests/test_wcsprm.py:
```diff
@@ -51,26 +51,26 @@ def test_cd():
     assert_array_equal(w.cd, [[1, 0], [0, 1]])
     del w.cd
-    assert w.has_cd() == False
+    assert w.has_cd() is False
```
Since you removed the `== True`, I would here and below probably do
```
assert not w.has_cd()
```

bsipocz replied:
I think this was an oversight from my side and those should have been ``is True`` as well to test their booleanness.

mhvk commented:
Some last comments (some of which may well be invalid). Otherwise all ready to go in as far as I'm concerned.

bsipocz commented:
@mhvk - Done. Does it need a changelog mention of adding E901,E902 to the config file? (I guess not, but better to ask.)

mhvk commented:
I don't think this needs a changelog entry, but cc @astrofrog in case.

mhvk commented:
I think I'll go ahead and merge. Thanks, @bsipocz!

astrofrog commented:
@bsipocz @mhvk - it looks like this introduced a bug in io.fits: https://github.com/astropy/astropy/issues/5807 - could you investigate to figure out which changes here caused that issue?
